### PR TITLE
Explicitly schedule telemetry pings

### LINKED ISSUE #230: Revisit Telemetry and get basic core ping sent

### PR DESCRIPTION
**Pull request by linuxwolf**
This PR explicitly schedules telemetry pings to be uploaded in two cases:

* When the Lockbox application moves to the background
* When a change to the `sendUsageData` setting is made

Fixes #230

_(**Required**: this reference (one or many) will be closed upon merge. Ideally it has the acceptance criteria and designs for features or fixes related to the work in this Pull Request.)_

## Testing and Review Notes

_(**Required**: steps to take to confirm this works as expected or other guidance for code, UX, and any other reviewers)_

* Open Lockbox
* Perform some activities that trigger telemetry events (for now it's just startup/foreground/background)
* Trigger an upload by toggling "send usage data" in Settings off/on
* Verify at the Telemetry Portal there are `Lockbox` events

## To Do

- add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
  - optional: consider adding instrumentation (integration/UI) tests
- consider running this branch in a debug simulator and check for memory leak notifications or any warnings
- ~request the "UX" team perform a design review (if/when applicable)~
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- ~check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI~

### CODE REVIEW

**Comment by linuxwolf** [comment]
Waiting some remote verification from @irrationalagent that any telemetry events are recorded.

**Comment by irrationalagent** [comment]
are you using `mozilla.lockbox` for the field `appName` ?

**Comment by linuxwolf** [comment]
> are you using `mozilla.lockbox` for the field `appName` ?

Sorry, I was mistaken ... the appName is set to `Lockbox`, although we can change that if desired/required.

**Comment by irrationalagent** [comment]
its good as is then, thanks

**Comment by irrationalagent** [comment]
btw no data yet, but it can take many hours. I'll post here in the morning if still nothing.

File: app/src/test/java/mozilla/lockbox/extensions/Observer+.kt [hunk 1]
```diff
@@ -11,4 +11,9 @@ import io.reactivex.observers.TestObserver
 fun <T> TestObserver<T>.assertLastValue(expectedValue: T) {
     val count = this.valueCount()
     this.assertValueAt(count - 1, expectedValue)
+}
+
+fun <T> TestObserver<T>.assertLastValueMatches(expectedPredicate: (value: T) -> Boolean) {
```
Comment (sashei) [inline app/src/test/java/mozilla/lockbox/extensions/Observer+.kt]: 😍 awesome operator!

File: app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt [hunk 2]
```diff
@@ -52,6 +53,9 @@ open class TelemetryWrapper {
     }
 
     open fun update(enabled: Boolean) {
+        // process any outstanding pings before settings change ...
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt]: 👍

File: app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt [hunk 3]
```diff
@@ -61,6 +65,14 @@ open class TelemetryWrapper {
     open fun recordEvent(event: TelemetryEvent) {
         subject?.queueEvent(event)
     }
+
+    open fun scheduleUpload() {
+        val telem = subject ?: return
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt]: nit, not related to this PR: I find the `subject` name a little confusing here, as we use it for the class under test in all the tests.

**Comment by irrationalagent** [comment]
It worked! We are receiving data now.

File: app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt (hunk 3 above)
Comment (linuxwolf) [reply app/src/main/java/mozilla/lockbox/store/TelemetryStore.kt]: I've no objections to changing it.  I think it was a pattern carried forward from somewhere.

**Comment by devinreams** [comment]
@linuxwolf Thanks for getting this working! 🥇

**Pull request merged**